pkg/controllers/jobset_controller.go

```diff
@@ -219,6 +222,63 @@ func (r *JobSetReconciler) getChildJobs(ctx context.Context, js *jobset.JobSet)
 	return &ownedJobs, nil
 }
 
+func (r *JobSetReconciler) calculateAndUpdateReplicatedJobsStatuses(ctx context.Context, js *jobset.JobSet, jobs *childJobs) {
```

Comment:
Could you refactor it so you can unit test a lot of the logic aside from the Status Update?

I see some cases I think you should target in a unit test but may be difficult to do as an integration test.  

I think modifying this function so it could return a (ReplicaedJobStatus, error) and you update the status.  That way you can at least unit test the majority of this logic.

Reply:
Yes, please do this. You could extract out everything up until the `r.Status().Update(...)` call into a separate function so that you can easily unit test the logic being used to set the values we are updating the status with. You should also return an error to trigger reconciliation as @kannon92 noted.

Pseudocode example:

```
func calculateReplicatedJobStatuses(...) {
  // all your logic for getting the statuses and add unit tests for this function
}

func (r* JobSetReconciler) calculateAndUpdateReplicatedJobsStatuses(...) (jobset.ReplicatedJobsStatus, error){
    status, err := calculateReplicatedJobStatuses(...)
    if err != nil {
        return nil, err
    }
    return r.Status().Update(...)
}
```